learn/shorts/functional-programming-python.qmd

```diff
@@ -0,0 +1,215 @@
+    imperative programming languages. We will focus on practical
+    applications of functional programming concepts in Python,
+    particularly the use of higher-order functions like `filter`, `map`,
+    and `reduce`. By the end of this lesson, learners will have hands-on
+    experience applying these techniques to write more concise and
+    expressive code.
+
+## Assumed knowledge of reader
+
+-   Basic programming knowledge
+-   Basic Python knowledge
+-   Basic understanding of how functions work
+
+## Necessary software
+
+-   Python editor of choice
+
+## Learning goal
+
+The goal of this lesson is for the reader to gain practical experience
+using functional programming techniques in Python. By the end, the
+reader should be able to answer the following questions:
+
+-   How can I use `filter`, `map`, and `reduce` in Python?
+-   What does the `lambda` keyword do in Python?
+
+## Take home messages
+
+-   Functional programming in Python allows for more concise and
+    readable code by leveraging built-in higher-order functions.
+-   The use of `filter`, `map`, and `reduce` enables efficient data
+    transformations without modifying original data structures.
+-   The `lambda` keyword allows for the creation of anonymous functions,
+    making functional programming more expressive and flexible.
+
+## Lesson content
+
+### How can I use Filter, Map, and Reduce in Python?
+
+#### **`filter`**
+
+This is a built-in function in Python that takes a function and an
+iterable as arguments and returns an iterator over the elements of the
+iterable for which the function returns `True`.
+
+##### Syntax:
+
+``` python
+filter(function, iterable)
+```
+
+The `function` should return `True` or `False` for each element in
+`iterable`. Only elements for which `function` returns `True` are
+included in the output.
+
+For example:
+
+``` python
+numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
+print(even_numbers)
+```
+
+In the example above, the `filter` function is used to filter out the
+even numbers from the list of numbers. The lambda function is used to
+define the condition for filtering, in this case, checking if the number
+produces a remainder of 0 when divided by 2. The `filter` function on
+its own returns what is called an iterator, so in order for us to see
+the result, we need to convert it to a list. We do the same below with
+the `map` function.
+
+#### **`map`**
+
+This is a built-in function in Python that takes a function and an
+iterable as arguments and returns an iterator over the results of
+applying the function to the elements of the iterable.
+
+##### Syntax:
+
+``` python
+map(function, iterable)
+```
+
+The `function` is applied to each element of `iterable`, and the results
+are returned as an iterator.
+
+For example:
+
+``` python
+numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+squared_numbers = list(map(lambda x: x ** 2, numbers))
```

Comment:
Here would be nice to show off *not* using lambda and instead use a function you create beforehand.